File: sdk/ts/src/wizard.ts
/**
 * Headless sequential wizard engine — TypeScript port of kit/wizard.
 */

// --- enums ---

export enum StepKind {
  TextInput = "text_input",
  Select = "select",
  Confirm = "confirm",
  MultiSelect = "multi_select",
  Action = "action",
  Summary = "summary",
}

export enum ErrorAction {
  Abort = 0,
  Retry = 1,
  Skip = 2,
}

// --- types ---

export interface Option {
  value: string;
  label: string;
  description?: string;
}

export interface Condition {
  key: string;
  pred: (value: unknown) => boolean;
}

export type ActionFn = (
  results: Record<string, unknown>,
) => void | Promise<void>;

export type ValidateTextFn = (value: string) => void;
export type ValidateChoiceFn = (value: string) => void;
export type ValidateChoicesFn = (values: string[]) => void;
export type FormatFn = (results: Record<string, unknown>) => string;

// --- Step ---

export class Step {
  key: string;
  kind: StepKind;
  label: string;
  description?: string;
  group?: string;
  required = false;
  defaultValue?: unknown;
  options?: Option[];
  validateText?: ValidateTextFn;
  validateChoice?: ValidateChoiceFn;
  validateChoices?: ValidateChoicesFn;
  when?: Condition;
  actionFn?: ActionFn;
  onError: ErrorAction = ErrorAction.Abort;
  formatFn?: FormatFn;

  constructor(init: {
    key: string;
    kind: StepKind;
    label: string;
    description?: string;
    group?: string;
    required?: boolean;
    defaultValue?: unknown;
    options?: Option[];
    validateText?: ValidateTextFn;
    validateChoice?: ValidateChoiceFn;
    validateChoices?: ValidateChoicesFn;
    when?: Condition;
    actionFn?: ActionFn;
    onError?: ErrorAction;
    formatFn?: FormatFn;
  }) {
    this.key = init.key;
    this.kind = init.kind;
    this.label = init.label;
    this.description = init.description;
    this.group = init.group;
    this.required = init.required ?? false;
    this.defaultValue = init.defaultValue;
    this.options = init.options;
    this.validateText = init.validateText;
    this.validateChoice = init.validateChoice;
    this.validateChoices = init.validateChoices;
    this.when = init.when;
    this.actionFn = init.actionFn;
    this.onError = init.onError ?? ErrorAction.Abort;
    this.formatFn = init.formatFn;
  }

  // --- chainable modifiers (return new Step) ---

  private clone(): Step {
    return new Step({ ...this });
  }

  withRequired(): Step {
    const s = this.clone();
    s.required = true;
    return s;
  }

  withDefault(v: unknown): Step {
    const s = this.clone();
    s.defaultValue = v;
    return s;
  }

  withGroup(g: string): Step {
    const s = this.clone();
    s.group = g;
    return s;
  }

  withDescription(d: string): Step {
    const s = this.clone();
    s.description = d;
    return s;
  }

  withValidateText(fn: ValidateTextFn): Step {
    const s = this.clone();
    s.validateText = fn;
    return s;
  }

  withValidateChoice(fn: ValidateChoiceFn): Step {
    const s = this.clone();
    s.validateChoice = fn;
    return s;
  }

  withValidateChoices(fn: ValidateChoicesFn): Step {
    const s = this.clone();
    s.validateChoices = fn;
    return s;
  }

  withWhen(key: string, pred: (value: unknown) => boolean): Step {
    const s = this.clone();
    s.when = { key, pred };
    return s;
  }

  withOnError(a: ErrorAction): Step {
    const s = this.clone();
    s.onError = a;
    return s;
  }

  withFormat(fn: FormatFn): Step {
    const s = this.clone();
    s.formatFn = fn;
    return s;
  }
}

// --- builders ---

export function textInput(key: string, label: string): Step {
  return new Step({ key, kind: StepKind.TextInput, label });
}

export function select(
  key: string,
  label: string,
  options: Option[],
): Step {
  return new Step({
    key,
    kind: StepKind.Select,
    label,
    options,
  });
}

export function confirm(key: string, label: string): Step {
  return new Step({ key, kind: StepKind.Confirm, label });
}

export function multiSelect(
  key: string,
  label: string,
  options: Option[],
): Step {
  return new Step({
    key,
    kind: StepKind.MultiSelect,
    label,
    options,
  });
}

export function action(
  key: string,
  label: string,
  fn: ActionFn,
): Step {
  return new Step({
    key,
    kind: StepKind.Action,
    label,
    actionFn: fn,
  });
}

export function summary(label: string): Step {
  return new Step({ key: "", kind: StepKind.Summary, label });
}

// --- errors ---

export class ValidationError extends Error {
  stepKey: string;

  constructor(stepKey: string, cause: Error | string) {
    const msg =
      typeof cause === "string" ? cause : cause.message;
    super(
      `validation failed for "${stepKey}": ${msg}`,
    );
    this.name = "ValidationError";
    this.stepKey = stepKey;
  }
}

export class ActionError extends Error {
  stepKey: string;
  cause: Error;
  action: ErrorAction;

  constructor(stepKey: string, cause: Error, act: ErrorAction) {
    super(`action "${stepKey}" failed: ${cause.message}`);
    this.name = "ActionError";
    this.stepKey = stepKey;
    this.cause = cause;
    this.action = act;
  }
}

export class ActionRequest {
  stepKey: string;
  run: ActionFn;

  constructor(stepKey: string, run: ActionFn) {
    this.stepKey = stepKey;
    this.run = run;
  }
}

// --- Wizard ---

export class Wizard {
  private steps: Step[];
  private _current = 0;
  private _results: Record<string, unknown> = {};
  private _done = false;
  private _dryRun = false;
  private _onComplete?: (
    results: Record<string, unknown>,
  ) => void;

  constructor(...steps: Step[]) {
    const seen = new Set<string>();
    for (let i = 0; i < steps.length; i++) {
      const s = steps[i];

      // auto-generate key for Summary
      if (s.kind === StepKind.Summary && s.key === "") {
        s.key = `__summary_${i}`;
      }

      if (s.key === "") {
        throw new Error(
          `step at index ${i}: key must not be empty`,
        );
      }

      if (seen.has(s.key)) {
        throw new Error(`duplicate step key "${s.key}"`);
      }
      seen.add(s.key);

      validateDefault(s);

      if (
        s.kind === StepKind.Action &&
        s.actionFn == null
      ) {
        throw new Error(
          `step "${s.key}": action kind must have actionFn`,
        );
      }

      if (
        (s.kind === StepKind.Select ||
          s.kind === StepKind.MultiSelect) &&
        (!s.options || s.options.length === 0)
      ) {
        throw new Error(
          `step "${s.key}": select/multi_select must have options`,
        );
      }
    }
    this.steps = steps;
  }

  current(): Step | null {
    while (this._current < this.steps.length) {
      const s = this.steps[this._current];
      if (
        s.when != null &&
        !s.when.pred(this._results[s.when.key])
      ) {
        delete this._results[s.key]; // clear stale
        this._current++;
        continue;
      }
      return s;
    }
    this._done = true;
    return null;
  }

  advance(
    value: unknown,
  ): [ActionRequest | null, ValidationError | null] {
    const s = this.current();
    if (s == null) return [null, null];

    switch (s.kind) {
      case StepKind.Action:
        return [
          new ActionRequest(s.key, s.actionFn!),
          null,
        ];

      case StepKind.Summary:
        this.doAdvance();
        return [null, null];

      case StepKind.TextInput:
      case StepKind.Select: {
        if (typeof value !== "string") {
          return [
            null,
            new ValidationError(s.key, "expected string"),
          ];
        }
        if (s.required && value === "") {
          return [
            null,
            new ValidationError(s.key, "required"),
          ];
        }
        if (
          s.kind === StepKind.TextInput &&
          s.validateText
        ) {
          try {
            s.validateText(value);
          } catch (e: unknown) {
            return [
              null,
              new ValidationError(
                s.key,
                e instanceof Error ? e : new Error(String(e)),
              ),
            ];
          }
        }
        if (s.kind === StepKind.Select) {
          if (!isValidOption(s.options!, value)) {
            return [
              null,
              new ValidationError(
                s.key,
                `invalid option "${value}"`,
              ),
            ];
          }
          if (s.validateChoice) {
            try {
              s.validateChoice(value);
            } catch (e: unknown) {
              return [
                null,
                new ValidationError(
                  s.key,
                  e instanceof Error
                    ? e
                    : new Error(String(e)),
                ),
              ];
            }
          }
        }
        this._results[s.key] = value;
        break;
      }

      case StepKind.Confirm: {
        if (typeof value !== "boolean") {
          return [
            null,
            new ValidationError(s.key, "expected bool"),
          ];
        }
        this._results[s.key] = value;
        break;
      }

      case StepKind.MultiSelect: {
        if (
          !Array.isArray(value) ||
          !value.every((v) => typeof v === "string")
        ) {
          return [
            null,
            new ValidationError(
              s.key,
              "expected string[]",
            ),
          ];
        }
        const ss = value as string[];
        if (s.required && ss.length === 0) {
          return [
            null,
            new ValidationError(s.key, "required"),
          ];
        }
        for (const v of ss) {
          if (!isValidOption(s.options!, v)) {
            return [
              null,
              new ValidationError(
                s.key,
                `invalid option "${v}"`,
              ),
            ];
          }
        }
        if (s.validateChoices) {
          try {
            s.validateChoices(ss);
          } catch (e: unknown) {
            return [
              null,
              new ValidationError(
                s.key,
                e instanceof Error
                  ? e
                  : new Error(String(e)),
              ),
            ];
          }
        }
        this._results[s.key] = ss;
        break;
      }
    }

    this.doAdvance();
    return [null, null];
  }

  private doAdvance(): void {
    this._current++;
    if (this.current() == null) {
      this._done = true;
    }
  }

  resolveAction(
    err: Error | null,
  ): ActionError | null {
    const s = this.current();
    if (s == null) return null;

    if (err == null) {
      this.doAdvance();
      return null;
    }

    switch (s.onError) {
      case ErrorAction.Abort:
        return new ActionError(
          s.key,
          err,
          ErrorAction.Abort,
        );
      case ErrorAction.Retry:
        return null;
      case ErrorAction.Skip:
        this.doAdvance();
        return null;
    }
    return new ActionError(s.key, err, s.onError);
  }

  back(): void {
    this._done = false;
    while (this._current > 0) {
      this._current--;
      const s = this.steps[this._current];
      if (
        s.when != null &&
        !s.when.pred(this._results[s.when.key])
      ) {
        continue;
      }
      delete this._results[s.key];
      return;
    }
  }

  results(): Record<string, unknown> {
    return { ...this._results };
  }

  done(): boolean {
    return this._done;
  }

  stepCount(): number {
    let n = 0;
    for (const s of this.steps) {
      if (
        s.when != null &&
        !s.when.pred(this._results[s.when.key])
      ) {
        continue;
      }
      n++;
    }
    return n;
  }

  stepIndex(): number {
    let idx = 0;
    for (
      let i = 0;
      i < this._current && i < this.steps.length;
      i++
    ) {
      const s = this.steps[i];
      if (
        s.when != null &&
        !s.when.pred(this._results[s.when.key])
      ) {
        continue;
      }
      idx++;
    }
    return idx;
  }

  setDryRun(v: boolean): void {
    this._dryRun = v;
  }

  dryRun(): boolean {
    return this._dryRun;
  }

  setOnComplete(
    fn: (results: Record<string, unknown>) => void,
  ): void {
    this._onComplete = fn;
  }

  complete(): void {
    if (this._dryRun || this._onComplete == null) return;
    this._onComplete(this._results);
  }
}

// --- validation helpers ---

function validateDefault(s: Step): void {
  if (s.defaultValue == null) return;

  switch (s.kind) {
    case StepKind.TextInput:
    case StepKind.Select:
      if (typeof s.defaultValue !== "string") {
        throw new Error(
          `step "${s.key}": default must be string for ${s.kind}`,
        );
      }
      break;
    case StepKind.Confirm:
      if (typeof s.defaultValue !== "boolean") {
        throw new Error(
          `step "${s.key}": default must be bool for confirm`,
        );
      }
      break;
    case StepKind.MultiSelect:
      if (
        !Array.isArray(s.defaultValue) ||
        !s.defaultValue.every(
          (v: unknown) => typeof v === "string",
        )
      ) {
        throw new Error(
          `step "${s.key}": default must be string[] for multi_select`,
        );
      }
      break;
  }
}

function isValidOption(opts: Option[], val: string): boolean {
  return opts.some((o) => o.value === val);
}

// --- result accessors ---

export function resultString(
  results: Record<string, unknown>,
  key: string,
): string {
  const v = results[key];
  return typeof v === "string" ? v : "";
}

export function resultBool(
  results: Record<string, unknown>,
  key: string,
): boolean {
  const v = results[key];
  return typeof v === "boolean" ? v : false;
}

export function resultStrings(
  results: Record<string, unknown>,
  key: string,
): string[] | null {
  const v = results[key];
  if (
    Array.isArray(v) &&
    v.every((x) => typeof x === "string")
  ) {
    return v as string[];
  }
  return null;
}

export function resultChoice(
  results: Record<string, unknown>,
  key: string,
): string {
  return resultString(results, key);
}

File: sdk/py/hop_top_kit/wizard.py
"""Headless sequential wizard engine -- Python port of kit/wizard."""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass, field
from enum import Enum
from typing import Any

# --- enums ---


class StepKind(Enum):
    TEXT_INPUT = "text_input"
    SELECT = "select"
    CONFIRM = "confirm"
    MULTI_SELECT = "multi_select"
    ACTION = "action"
    SUMMARY = "summary"


class ErrorAction(Enum):
    ABORT = 0
    RETRY = 1
    SKIP = 2


# --- types ---


@dataclass
class Option:
    value: str
    label: str
    description: str = ""


@dataclass
class Condition:
    key: str
    pred: Callable[[Any], bool]


# --- errors ---


class ValidationError(Exception):
    def __init__(self, step_key: str, msg: str):
        super().__init__(f'validation failed for "{step_key}": {msg}')
        self.step_key = step_key


class ActionError(Exception):
    def __init__(
        self,
        step_key: str,
        cause: Exception,
        action: ErrorAction,
    ):
        super().__init__(f'action "{step_key}" failed: {cause}')
        self.step_key = step_key
        self.cause = cause
        self.action = action


class ActionRequest:
    def __init__(
        self,
        step_key: str,
        run: Callable[[dict[str, Any]], Any],
    ):
        self.step_key = step_key
        self.run = run


# --- Step ---


@dataclass
class Step:
    key: str
    kind: StepKind
    label: str
    description: str = ""
    group: str = ""
    required: bool = False
    default_value: Any = None
    options: list[Option] = field(default_factory=list)
    validate_text_fn: Callable[[str], None] | None = None
    validate_choice_fn: Callable[[str], None] | None = None
    validate_choices_fn: Callable[[list[str]], None] | None = None
    when: Condition | None = None
    action_fn: Callable[[dict[str, Any]], Any] | None = None
    on_error: ErrorAction = ErrorAction.ABORT
    format_fn: Callable[[dict[str, Any]], str] | None = None

    # --- chainable modifiers (return new Step) ---

    def _clone(self, **kwargs: Any) -> Step:
        import dataclasses

        return dataclasses.replace(self, **kwargs)

    def with_required(self) -> Step:
        return self._clone(required=True)

    def with_default(self, v: Any) -> Step:
        return self._clone(default_value=v)

    def with_group(self, g: str) -> Step:
        return self._clone(group=g)

    def with_description(self, d: str) -> Step:
        return self._clone(description=d)

    def with_validate_text(self, fn: Callable[[str], None]) -> Step:
        return self._clone(validate_text_fn=fn)

    def with_validate_choice(self, fn: Callable[[str], None]) -> Step:
        return self._clone(validate_choice_fn=fn)

    def with_validate_choices(self, fn: Callable[[list[str]], None]) -> Step:
        return self._clone(validate_choices_fn=fn)

    def with_when(
        self,
        key: str,
        pred: Callable[[Any], bool],
    ) -> Step:
        return self._clone(when=Condition(key=key, pred=pred))

    def with_on_error(self, a: ErrorAction) -> Step:
        return self._clone(on_error=a)

    def with_format(self, fn: Callable[[dict[str, Any]], str]) -> Step:
        return self._clone(format_fn=fn)


# --- builders ---


def _to_option(o: dict[str, str] | Option) -> Option:
    if isinstance(o, Option):
        return o
    return Option(
        value=o["value"],
        label=o["label"],
        description=o.get("description", ""),
    )


def text_input(key: str, label: str) -> Step:
    return Step(key=key, kind=StepKind.TEXT_INPUT, label=label)


def select(
    key: str,
    label: str,
    options: list[dict[str, str] | Option],
) -> Step:
    return Step(
        key=key,
        kind=StepKind.SELECT,
        label=label,
        options=[_to_option(o) for o in options],
    )


def confirm(key: str, label: str) -> Step:
    return Step(key=key, kind=StepKind.CONFIRM, label=label)


def multi_select(
    key: str,
    label: str,
    options: list[dict[str, str] | Option],
) -> Step:
    return Step(
        key=key,
        kind=StepKind.MULTI_SELECT,
        label=label,
        options=[_to_option(o) for o in options],
    )


def action(
    key: str,
    label: str,
    fn: Callable[[dict[str, Any]], Any],
) -> Step:
    return Step(
        key=key,
        kind=StepKind.ACTION,
        label=label,
        action_fn=fn,
    )


def summary(label: str) -> Step:
    return Step(key="", kind=StepKind.SUMMARY, label=label)


# --- validation helpers ---


def _validate_default(s: Step) -> None:
    if s.default_value is None:
        return
    k = s.kind
    if k in (StepKind.TEXT_INPUT, StepKind.SELECT):
        if not isinstance(s.default_value, str):
            raise ValueError(f'step "{s.key}": default must be string for {k.value}')
    elif k == StepKind.CONFIRM:
        if not isinstance(s.default_value, bool):
            raise ValueError(f'step "{s.key}": default must be bool for confirm')
    elif k == StepKind.MULTI_SELECT and not (
        isinstance(s.default_value, list) and all(isinstance(v, str) for v in s.default_value)
    ):
        raise ValueError(f'step "{s.key}": default must be list[str] for multi_select')


def _is_valid_option(opts: list[Option], val: str) -> bool:
    return any(o.value == val for o in opts)


# --- Wizard ---


class Wizard:
    def __init__(self, *steps: Step):
        self._steps = list(steps)
        self._current = 0
        self._results: dict[str, Any] = {}
        self._done = False
        self._dry_run = False
        self._on_complete: Callable[[dict[str, Any]], Any] | None = None

        seen: set[str] = set()
        for i, s in enumerate(self._steps):
            # auto-generate key for Summary
            if s.kind == StepKind.SUMMARY and s.key == "":
                s.key = f"__summary_{i}"

            if s.key == "":
                raise ValueError(f"step at index {i}: key must not be empty")

            if s.key in seen:
                raise ValueError(f'duplicate step key "{s.key}"')
            seen.add(s.key)

            _validate_default(s)

            if s.kind == StepKind.ACTION and s.action_fn is None:
                raise ValueError(f'step "{s.key}": action kind must have action_fn')

            if (
                s.kind
                in (
                    StepKind.SELECT,
                    StepKind.MULTI_SELECT,
                )
                and not s.options
            ):
                raise ValueError(f'step "{s.key}": select/multi_select must have options')

    def current(self) -> Step | None:
        while self._current < len(self._steps):
            s = self._steps[self._current]
            if s.when is not None and not s.when.pred(self._results.get(s.when.key)):
                self._results.pop(s.key, None)  # clear stale
                self._current += 1
                continue
            return s
        self._done = True
        return None

    def advance(self, value: Any) -> tuple[ActionRequest | None, ValidationError | None]:
        s = self.current()
        if s is None:
            return None, None

        k = s.kind

        if k == StepKind.ACTION:
            return (
                ActionRequest(s.key, s.action_fn),
                None,
            )

        if k == StepKind.SUMMARY:
            self._do_advance()
            return None, None

        if k in (StepKind.TEXT_INPUT, StepKind.SELECT):
            if not isinstance(value, str):
                return None, ValidationError(s.key, "expected string")
            if s.required and value == "":
                return None, ValidationError(s.key, "required")
            if k == StepKind.TEXT_INPUT and s.validate_text_fn is not None:
                try:
                    s.validate_text_fn(value)
                except Exception as e:
                    return None, ValidationError(s.key, str(e))
            if k == StepKind.SELECT:
                if not _is_valid_option(s.options, value):
                    return None, ValidationError(
                        s.key,
                        f'invalid option "{value}"',
                    )
                if s.validate_choice_fn is not None:
                    try:
                        s.validate_choice_fn(value)
                    except Exception as e:
                        return None, ValidationError(s.key, str(e))
            self._results[s.key] = value

        elif k == StepKind.CONFIRM:
            if not isinstance(value, bool):
                return None, ValidationError(s.key, "expected bool")
            self._results[s.key] = value

        elif k == StepKind.MULTI_SELECT:
            if not isinstance(value, list) or not all(isinstance(v, str) for v in value):
                return None, ValidationError(s.key, "expected list[str]")
            if s.required and len(value) == 0:
                return None, ValidationError(s.key, "required")
            for v in value:
                if not _is_valid_option(s.options, v):
                    return None, ValidationError(
                        s.key,
                        f'invalid option "{v}"',
                    )
            if s.validate_choices_fn is not None:
                try:
                    s.validate_choices_fn(value)
                except Exception as e:
                    return None, ValidationError(s.key, str(e))
            self._results[s.key] = value

        self._do_advance()
        return None, None

    def _do_advance(self) -> None:
        self._current += 1
        if self.current() is None:
            self._done = True

    def resolve_action(self, err: Exception | None) -> ActionError | None:
        s = self.current()
        if s is None:
            return None

        if err is None:
            self._do_advance()
            return None

        if s.on_error == ErrorAction.ABORT:
            return ActionError(s.key, err, ErrorAction.ABORT)
        if s.on_error == ErrorAction.RETRY:
            return None
        if s.on_error == ErrorAction.SKIP:
            self._do_advance()
            return None

        return ActionError(s.key, err, s.on_error)

    def back(self) -> None:
        self._done = False
        while self._current > 0:
            self._current -= 1
            s = self._steps[self._current]
            if s.when is not None and not s.when.pred(self._results.get(s.when.key)):
                continue
            self._results.pop(s.key, None)
            return

    def results(self) -> dict[str, Any]:
        return dict(self._results)

    def done(self) -> bool:
        return self._done

    def step_count(self) -> int:
        n = 0
        for s in self._steps:
            if s.when is not None and not s.when.pred(self._results.get(s.when.key)):
                continue
            n += 1
        return n

    def step_index(self) -> int:
        idx = 0
        for i in range(min(self._current, len(self._steps))):
            s = self._steps[i]
            if s.when is not None and not s.when.pred(self._results.get(s.when.key)):
                continue
            idx += 1
        return idx

    def set_dry_run(self, v: bool) -> None:
        self._dry_run = v

    def dry_run(self) -> bool:
        return self._dry_run

    def set_on_complete(self, fn: Callable[[dict[str, Any]], Any]) -> None:
        self._on_complete = fn

    def complete(self) -> None:
        if self._dry_run or self._on_complete is None:
            return
        self._on_complete(self._results)


# --- result accessors ---


def result_string(results: dict[str, Any], key: str) -> str:
    v = results.get(key)
    return v if isinstance(v, str) else ""


def result_bool(results: dict[str, Any], key: str) -> bool:
    v = results.get(key)
    return v if isinstance(v, bool) else False


def result_strings(results: dict[str, Any], key: str) -> list[str] | None:
    v = results.get(key)
    if isinstance(v, list) and all(isinstance(x, str) for x in v):
        return v
    return None


def result_choice(results: dict[str, Any], key: str) -> str:
    return result_string(results, key)
